### PR TITLE
refactor overlap calculation code

### PR DESCRIPTION
**Pull request by tst-basilhuffman**
Initial commit, will build and cancel sample imagery but integration tests fail

### CODE REVIEW

File: docs/geedocs/5.2.5/answer/176738.html [hunk 1]
```diff
@@ -257,6 +257,10 @@ <h2 id="cache_configuration">Cache configuration</h2>
 Note that the asset version cache is generally larger than the asset cache because it may contain multiple versions of the same asset.</p>
 
 Also included within <code>misc.xml</code> is an experimental option, which allows for the user to disable cache purging while projects build. This is controlled by <code>DisablePacketLevelVersionCachePurge</code> and its default value is <code>false</code>.
+
+<h2 id="overlap_calculation">Overlap Calculation</h2>
```
Comment (tst-lsavoie) [inline docs/geedocs/5.2.5/answer/176738.html]: Thanks for updating the docs. These changes are going to end up in 5.3.1, so that's where the documentation should go, not 5.2.5. The only way to do that right now is to have a separate PR for the docs.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 2]
```diff
@@ -996,28 +989,30 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
...
+      extraArg.genInfo.push_back(&genInfo[idx]);
+    }
 
+      // reuse or make a new PacketGenAssetVersion
+      MutableAssetVersionD packgen =
+        PacketGenFactory::ReuseOrMakeAndUpdate(genInfo[i].assetname,
+                                               type,
+                                               khMetaData(),
+                                               packetGenConfig,
+                                               extraArg,
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Not a huge deal, but I think the idea behind `extraArg` is to include any extra arguments that are specific to this type of asset. I think it would be cleaner to move these two new arguments into extraArg. That way you wouldn't have to change the function signature.

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src [hunk 3]
```diff
@@ -1038,6 +1035,25 @@ uint32 RasterProjectAssetVersionImplD::BuildIndex(
   packgenkids.reserve(numInsets);
   packgenassets.reserve(numInsets);
   packgenvers.reserve(numInsets);
+  std::vector<uint> totalNeededIndexes;
+
+// calculate total overlap before entering here
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: Looks like this is debug code that was not intended to be permanent.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp [hunk 4]
```diff
@@ -207,6 +207,12 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
 
   // populate list - inital coverage estimate from inset degExtents
   genInfos.reserve(insetInfos.size());
+  /*
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp]: Looks like this was not intended to be permanent.

**Comment by tst-lsavoie** [comment]
I put a little thought into this and I think this is pretty close to what we need. The overall idea is that we want to make it easy to swap out overlap algorithms. Separating the overlap calls into a function is a step towards accomplishing that. My original thought was that calculating the overlap all at once would be the most generic method since we can simply replace that all-at-once calculation with a different all at once calculation. After looking at this I think there's a simpler way.

In general, this is what needs to happen:

1. Perform some general calculations to get partial overlap data (this happens in BuildIndex).
1. Pass this partial data down to UpdateChildren.
1. Calculate the exact overlap for each level in UpdateChildren.

I think you can accomplish this with just a couple of function name changes. We could call the function called from BuildIndex something like `PreprocessOverlapForInset` and the function called from UpdateChildren something like `GetOverlapForLevel`. For this version of the code they would just call the same CalculateOverlap function that you've already written, but when we start working on overlap calculations they could do different things. That also eliminates the need for a flag since the calculations will be the same for this version, they'll just be organized differently.

To take it a step farther you could create an overlap class and put the two functions into it, then just pass the whole object from BuildIndex to UpdateChildren.

In summary, here's what I think needs to be done to finish up this ticket:

1. Put the two overlap calculation calls inside wrapper functions that describe what they're doing (as I described above). As I mentioned, you could create a class for this.
1. Remove the changes to the function signature for `ReuseOrMakeAndUpdate`.
1. Remove the new flag in MiscConfig.
1. Remove debugging code as noted in comments.
1. Move the doc changes to a separate PR so they can be in the 5.3.1 docs.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 5]
```diff
@@ -85,9 +88,11 @@ PacketGenFactory::ReuseOrMakeAndUpdateSubAsset(
 
 template <typename ProductAssetVersion>
 void PacketGenAssetVersionImplD::UpdateChildren(
-    const PacketGenExtraUpdateArg<ProductAssetVersion> &extra) {
+    const PacketGenExtraUpdateArg<ProductAssetVersion> &extra,
+    bool skip, const std::vector<uint>& _neededIndexes) {
```
Comment (tst-sbayard) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: I feel like `skip` may be ambiguous for anyone looking at the function signature. Suggest renaming here and other uses to something that makes it clear what is being skipped.

Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: this is getting removed

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 6]
```diff
@@ -122,21 +124,23 @@ void PacketGenAssetVersionImplD::UpdateChildren(
...
+
+    // there are situations where this can happen and must be accounted for
+    if (calc == nullptr)
+    {
+        calc.reset(new OverlapCalculator<ProductAssetVersion>(env));
+        calc->PreprocessForInset();
     }
 
+    neededIndexes = std::move(calc->GetOverlapForLevel());
+    //neededIndexes = std::move(CalculateOverlap(env));
```
Comment (tst-sbayard) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Should remove this commented-out code.

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src [hunk 7]
```diff
@@ -1057,19 +1054,15 @@ uint32 RasterProjectAssetVersionImplD::BuildIndex(
...
-    }
+    overlapEnvelope<RasterProductAssetVersion> env (
+        type, genInfo[i].coverage, insetInfoPtrs, i,
+        beginMinifyLevel, endMinifyLevel);
+
+    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
+        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
+    calc->PreprocessForInset();
+    neededIndexes = std::move(calc->GetOverlapForLevel());
+    //neededIndexes = std::move(CalculateOverlap(env));
```
Comment (tst-sbayard) [inline earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: Remove commented-out code.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h [hunk 8]
```diff
@@ -168,6 +168,55 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
...
+    void PreprocessForInset()
+    {
+        neededIndexes.clear();
+        CalculateOverlap();
+    }
+
+    std::vector<uint> GetOverlapForLevel() { return neededIndexes; }
+
+};
+
```
Comment (tst-sbayard) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h]: Should remove the commented-out code below if it's not needed.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 9]
```diff
@@ -33,6 +33,7 @@
 #include "autoingest/plugins/RasterDBRootAsset.h"
 #include "autoingest/plugins/MercatorRasterProductAsset.h"
 #include "autoingest/plugins/MapLayerJSAsset.h"
+#include "autoingest/MiscConfig.h"
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Minor: I think this can be removed.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 10]
```diff
@@ -122,21 +124,23 @@ void PacketGenAssetVersionImplD::UpdateChildren(
...
+                         config.beginMinifyLevel, config.endMinifyLevel);
+
+    // there are situations where this can happen and must be accounted for
+    if (calc == nullptr)
+    {
+        calc.reset(new OverlapCalculator<ProductAssetVersion>(env));
+        calc->PreprocessForInset();
     }
 
+    neededIndexes = std::move(calc->GetOverlapForLevel());
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: I believe the level needs to be passed as a parameter here. If I'm not mistaken that's one of the important pieces of information for figuring out the overlap for each level.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 11]
```diff
@@ -942,19 +943,15 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
...
-                              neededIndexes,
-                              beginMinifyLevel,
-                              endMinifyLevel);
-    }
+
+    overlapEnvelope<MercatorRasterProductAssetVersion> env (
+        type, genInfo[i].coverage, insetInfoPtrs, i,
+        beginMinifyLevel, endMinifyLevel);
+
+    std::shared_ptr<OverlapCalculator<MercatorRasterProductAssetVersion>> calc =
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Very minor: I suspect this could be a unique_ptr instead of a shared_ptr, which has slightly better performance.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 12]
```diff
@@ -996,28 +992,29 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
...
+    }
 
+      // reuse or make a new PacketGenAssetVersion
+      MutableAssetVersionD packgen =
+        PacketGenFactory::ReuseOrMakeAndUpdate(genInfo[i].assetname,
+                                               type,
+                                               khMetaData(),
+                                               packetGenConfig,
+                                               extraArg,
+                                               calc);
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: I'd really like to see calc become a member of extraArg. That way you don't need an extra parameter.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src (hunk 10 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: unless i misunderstand it, is already being covered within khinsetcoverage, but i am passing it down either way

File: earth_enterprise/src/fusion/autoingest/AssetGenDaemonH.pl [hunk 13]
```diff
@@ -98,7 +103,7 @@ ()
 
 #include <$header>
 #include <sysman/AssetD.h>
-
+#include <memory>
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/AssetGenDaemonH.pl]: Minor: I think you can remove this include now.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 14]
```diff
@@ -868,9 +865,10 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
   if (type == AssetDefs::Imagery) {
     // first we know that we don't want skip tiles during minification
     // so set each inset to start after that
-    for (uint i = 0; i < genInfo.size(); ++i) {
-      genInfo[i].beginSkipTransparentLevel =
-        std::max(endMinifyLevel, genInfo[i].coverage.beginLevel());
+
+    for (auto& i : genInfo) {
+        i.beginSkipTransparentLevel =
+        std::max(endMinifyLevel, i.coverage.beginLevel());
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Minor: would be good to have indentation for the second line.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp [hunk 15]
```diff
@@ -330,6 +331,31 @@ void CalcPacketGenInfo<InsetInfo<MercatorRasterProductAssetVersion> > (
...
+        FindNeededImageryInsets(env.gencov,
+                                env.insets,
+                                env.numInsets,
+                                neededIndexes,
+                                env.beginMinifyLevel,
+                                env.endMinifyLevel);
+    }
+    else
+    {
+        FindNeededImageryInsets(env.gencov,
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp]: Shouldn't this second call be `FindNeededTerrainInsets`?

**Comment by tst-lsavoie** [comment]
This is performance critical code, so I'm being extra picky.

Looking at the old code, this is roughly what happens:
```
Inside BuildIndex
For each genInfo:
  Call FindNeededImageryInsets // Returns all insets that overlap with the current inset at any level
  Call PacketGenFactory::ReuseOrMakeAndUpdate which eventually calls UpdateChildren
  
Inside UpdateChildren
For each level:
  Call FindNeededImageryInsets // Returns all insets that overlap with the current inset at a specific level
```

In the new code, this is what happens:
```
Inside BuildIndex
For each genInfo:
  Call calc->PreprocessForInset
    Which calls FindNeededImageryInsets // Returns all insets that overlap with the current inset at any level
  Call PacketGenFactory::ReuseOrMakeAndUpdate which eventually calls UpdateChildren

Inside UpdateChildren
For each level:
  Call calc->GetOverlapForLevel (does not call FindNeededImageryInsets)
```

The code in BuildIndex does the same thing in both versions, but the second call to FindNeededImageryInsets in UpdateChildren is missing in the new version. If I understand correctly, that means we'll have extra insets and do unnecessary processing. The ultimate goal here is to refactor the code such that it works exactly the same but it's easy to swap out different methods of calculating overlap.

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src [hunk 16]
```diff
@@ -1057,19 +1054,14 @@ uint32 RasterProjectAssetVersionImplD::BuildIndex(
...
-    }
+    overlapEnvelope<RasterProductAssetVersion> env (
+        type, genInfo[i].coverage, insetInfoPtrs, i,
+        beginMinifyLevel, endMinifyLevel, i);
+
+    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
+        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
+    calc->PreprocessForInset();
+    neededIndexes = std::move(calc->GetOverlapForLevel());
     neededIndexes.push_back(i); // each inset always intersects itself
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: I think there may be a problem here. Since we're adding the inset outside of the calc object the version of neededIndexes that's inside the object will not have the extra element.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src (hunk 14 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: done

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp (hunk 15 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp]: fat fingered. fixed

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src (hunk 16 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: fixed

**Comment by tst-basilhuffman** [comment]
> This is performance critical code, so I'm being extra picky.
> 
> Looking at the old code, this is roughly what happens:
> 
> ```
> Inside BuildIndex
> For each genInfo:
>   Call FindNeededImageryInsets // Returns all insets that overlap with the current inset at any level
>   Call PacketGenFactory::ReuseOrMakeAndUpdate which eventually calls UpdateChildren
>   
> Inside UpdateChildren
> For each level:
>   Call FindNeededImageryInsets // Returns all insets that overlap with the current inset at a specific level
> ```
> 
> In the new code, this is what happens:
> 
> ```
> Inside BuildIndex
> For each genInfo:
>   Call calc->PreprocessForInset
>     Which calls FindNeededImageryInsets // Returns all insets that overlap with the current inset at any level
>   Call PacketGenFactory::ReuseOrMakeAndUpdate which eventually calls UpdateChildren
> 
> Inside UpdateChildren
> For each level:
>   Call calc->GetOverlapForLevel (does not call FindNeededImageryInsets)
> ```
> 
> The code in BuildIndex does the same thing in both versions, but the second call to FindNeededImageryInsets in UpdateChildren is missing in the new version. If I understand correctly, that means we'll have extra insets and do unnecessary processing. The ultimate goal here is to refactor the code such that it works exactly the same but it's easy to swap out different methods of calculating overlap.

So, for now, UpdateChildren **should** be making that call to `FindNeededImageryInsets`?

**Comment by tst-lsavoie** [comment]
Yes.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 17]
```diff
@@ -120,22 +120,19 @@ void PacketGenAssetVersionImplD::UpdateChildren(
...
-                              neededIndexes,
-                              config.beginMinifyLevel,
-                              config.endMinifyLevel);
-    }
+    auto gencov = khInsetCoverage(levCov);
+    //overlapEnvelope<ProductAssetVersion> env(type, gencov,
+    //                     extra.insetInfo, extra.insetInfo.size(),
+    //                     config.beginMinifyLevel, config.endMinifyLevel, level);
+
+    extra.calc->env.gencov = gencov;
```
Comment (tst-sbayard) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Since `extra.calc->env.gencov` is a `khInsetCoverage&`, it seems like `extra.calc->env.gencov = gencov;` is going to end up modifying the object that `extra.calc->env.gencov` references ( `genInfo[i].coverage` from *RasterProject.src) instead of modifying the reference itself.
Not sure if that's what we want to do.

Comment (tst-sbayard) [reply earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Resolved by latest commit.

File: earth_enterprise/src/common/khInsetCoverage.h [hunk 18]
```diff
@@ -39,6 +39,7 @@ class khInsetCoverage {
   inline uint beginLevel(void) const { return beginLevel_; }
   inline uint endLevel(void)   const { return endLevel_; }
   inline uint numLevels(void)  const { return endLevel() - beginLevel(); }
+  inline uint numVLvls(void)  const { return vec0Level; }
```
[the remark this answers is not in the document]
Comment (tst-sbayard) [reply earth_enterprise/src/common/khInsetCoverage.h]: Leaving in for now as it may be useful for upcoming chanes.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 19]
```diff
@@ -868,9 +865,10 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
   if (type == AssetDefs::Imagery) {
     // first we know that we don't want skip tiles during minification
     // so set each inset to start after that
-    for (uint i = 0; i < genInfo.size(); ++i) {
-      genInfo[i].beginSkipTransparentLevel =
-        std::max(endMinifyLevel, genInfo[i].coverage.beginLevel());
+
+    for (auto& i : genInfo) {
+        i.beginSkipTransparentLevel =
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Minor: would be good to reduce this indentation by 2 spaces so that the next line is indented more than this one.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 20]
```diff
@@ -942,19 +940,15 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
...
-                              neededIndexes,
-                              beginMinifyLevel,
-                              endMinifyLevel);
-    }
+
+    overlapEnvelope<MercatorRasterProductAssetVersion> env (
+        type, genInfo[i].coverage, insetInfoPtrs, i,
+        beginMinifyLevel, endMinifyLevel, i);
+
+    std::shared_ptr<OverlapCalculator<MercatorRasterProductAssetVersion>> calc =
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Minor: thinking about this a bit more: do you need a shared_ptr at all here? I'm not sure what the shared_ptr is getting us since we don't have lots of parts of the code using this over and over. I suspect just passing around the object (rather than a shared pointer to the object) would work fine.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 21]
```diff
@@ -120,22 +120,19 @@ void PacketGenAssetVersionImplD::UpdateChildren(
...
-                              config.endMinifyLevel);
-    } else {
-      FindNeededTerrainInsets(khInsetCoverage(levCov),
-                              extra.insetInfo, numInsets,
-                              neededIndexes,
-                              config.beginMinifyLevel,
-                              config.endMinifyLevel);
-    }
+    auto gencov = khInsetCoverage(levCov);
+    //overlapEnvelope<ProductAssetVersion> env(type, gencov,
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Remove commented out code.

File: earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src [hunk 22]
```diff
@@ -996,27 +989,27 @@ uint32 MercatorRasterProjectAssetVersionImplD::BuildIndex(
...
     }
 
-    // reuse or make a new PacketGenAssetVersion
-    MutableAssetVersionD packgen =
-      PacketGenFactory::ReuseOrMakeAndUpdate(genInfo[i].assetname,
-                                             type,
-                                             khMetaData(),
-                                             packetGenConfig,
-                                             extraArg);
+      // reuse or make a new PacketGenAssetVersion
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/MercatorRasterProject.src]: Minor: I think the extra indentation was added here by mistake.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 23]
```diff
@@ -120,22 +120,19 @@ void PacketGenAssetVersionImplD::UpdateChildren(
     prevTime = curTime;
 
     // Determine which insets intersect us at this level
-    const khLevelCoverage levCov(config.coverage.levelCoverage(level));
+    khLevelCoverage levCov(config.coverage.levelCoverage(level));
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Any reason this is not const anymore?

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src [hunk 24]
```diff
@@ -1132,20 +1123,23 @@ uint32 RasterProjectAssetVersionImplD::BuildIndex(
...
 
-    // reuse or make a new PacketGenAssetVersion
-    MutableAssetVersionD packgen =
-      PacketGenFactory::ReuseOrMakeAndUpdate(genInfo[i].assetname,
-                                             type,
-                                             khMetaData(),
-                                             packetGenConfig,
-                                             extraArg);
+    // reuse or make a new PacketGenAssetVersion --> update children
+    // refactor this method call if wanting to do all calculations at once
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: Minor: I think we can remove this comment.

Also minor: I think the extra indentation was added below by accident.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp (hunk 15 above)
Comment (tst-lsavoie) [reply earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.cpp]: I'm not sure what happened but this still calls the imagery version, not the terrain version.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h [hunk 25]
```diff
@@ -168,6 +168,60 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
...
+    overlapEnvelope<ProductAssetVersion>& env;
+
+    void PreprocessForInset()
+    {
+        neededIndexes.clear();
+        CalculateOverlap();
+    }
+
+    std::vector<uint> GetOverlapForLevel(bool recalculate = false)
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h]: Minor: you could get rid of the boolean by just returning neededIndexes from PreprocessForInset and not calling GetOverlapForLevel at all from BuildIndex.

File: earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src [hunk 26]
```diff
@@ -1057,19 +1054,14 @@ uint32 RasterProjectAssetVersionImplD::BuildIndex(
...
-                              endMinifyLevel);
-    } else {
-      FindNeededTerrainInsets(genInfo[i].coverage,
-                              insetInfoPtrs, i,
-                              neededIndexes,
-                              beginMinifyLevel,
-                              endMinifyLevel);
-    }
+    overlapEnvelope<RasterProductAssetVersion> env (
+        type, genInfo[i].coverage, insetInfoPtrs, i,
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/RasterProject.src]: Passing `i` twice here seems unnecessary. It would be nice to only do it once.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h [hunk 27]
```diff
@@ -168,6 +168,60 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
...
+                    numInsets(_numInsets),
+                    beginMinifyLevel(_beginMinifyLevel), endMinifyLevel(_endMinifyLevel),
+                    level(_level) {}
+};
+
+template <typename ProductAssetVersion>
+class OverlapCalculator
+{
+private:
+    std::vector<uint> neededIndexes;
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h]: I think we can get away without storing neededIndexes permanently in this object for now since we're clearing it every time we use it. That will save us some memory, which is an important concern right now. Here's my recommendation.

In the `OverlapCalculator` class, remove the `neededIndexes` member variable.

Change `PreprocessForInset` from
```
    void PreprocessForInset()
    {
        neededIndexes.clear();
        CalculateOverlap();
    }
```
to
```
    std::vector<uint> PreprocessForInset()
    {
        std::vector<uint> neededIndexes;
        CalculateOverlap(neededIndexes);
        return neededIndexes;
    }
```

In BuildIndex, change:
```
    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
    calc->PreprocessForInset();
    neededIndexes = std::move(calc->GetOverlapForLevel());
```
to
```
    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
    neededIndexes = std::move(calc->PreprocessForInset());
```

There are a few other changes that will be necessary to make this work but I think those are the main ones. This will also allow you to get rid of the boolean parameter to GetOverlapForLevel and just always recalculate when that is called.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src (hunk 17 above)
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: Making a member variable public and modifying it external to the class is not a great pattern. It would be best to handle this another way. Maybe gencov shouldn't be part of the envelope but should be passed to the calls that need it.

File: earth_enterprise/src/fusion/autoingest/AssetGenDaemonH.pl (hunk 13 above)
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/AssetGenDaemonH.pl]: I don't think this include is needed.

File: earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src [hunk 28]
```diff
@@ -120,22 +120,15 @@ void PacketGenAssetVersionImplD::UpdateChildren(
...
-                              extra.insetInfo, numInsets,
-                              neededIndexes,
-                              config.beginMinifyLevel,
-                              config.endMinifyLevel);
-    }
+    auto gencov = khInsetCoverage(levCov);
+
+    extra.calc->setBeginMinifyLevel(config.beginMinifyLevel);
+    extra.calc->setEndMinifyLevel(config.endMinifyLevel);
+    neededIndexes = std::move(extra.calc->GetOverlapForLevel(gencov));
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/plugins.src/PacketGen.src]: I just realized that the `extra.insetInfo` variable is different here than it is in BuildIndex. I think this will work but it may take longer. I think the most efficient way to do this is to pass in the entire envelope each time to make sure that we get the right inputs.

File: earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h [hunk 29]
```diff
@@ -168,6 +168,60 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
...
+{
+private:
+    void CalculateOverlap(std::vector<uint>&, const khInsetCoverage&);
+    overlapEnvelope<ProductAssetVersion> env;
+
+public:
+    OverlapCalculator(const overlapEnvelope<ProductAssetVersion>& _env) :
+        env(_env) {}
+
+    void setBeginMinifyLevel(uint _begin) { env.beginMinifyLevel = _begin; }
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h]: Minor: I don't think you need these two functions any more.

**Comment by tst-lsavoie** [comment]
I just realized this was targeted to 5.2.5.1. I've retargeted it to master (where we want to merge it), but that has created some merge conflicts.

**Comment by tst-lsavoie** [comment]
I'm fine with including my commits on this PR.
